micro_http/src/connection.rs

```diff
@@ -0,0 +1,675 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::collections::VecDeque;
+use std::io::{Read, Write};
+
+use common::ascii::{CR, CRLF_LEN, LF};
+use common::Body;
+pub use common::{ConnectionError, RequestError};
+use headers::Headers;
+use request::{find, Request, RequestLine};
+use response::{Response, StatusCode};
+
+const BUFFER_SIZE: usize = 1024;
+
+/// Describes the state machine of an HTTP connection.
+pub enum ConnectionState {
+    WaitingForRequestLine,
+    WaitingForHeaders,
+    WaitingForBody,
+    RequestReady,
+}
+
+/// A wrapper over a HTTP Connection.
+pub struct HttpConnection<T> {
+    pending_request: Option<Request>,
+    stream: T,
+    state: ConnectionState,
+    buffer: [u8; BUFFER_SIZE],
+    current_read_start: usize,
+    body_vec: Vec<u8>,
+    body_bytes_to_be_read: i32,
+    parsed_requests: VecDeque<Request>,
+    response_queue: VecDeque<Response>,
+    response_buffer: Option<Vec<u8>>,
+}
+
+impl<T: Read + Write> HttpConnection<T> {
+    /// Creates a new default connection.
+    pub fn new(stream: T) -> Self {
+        HttpConnection {
+            pending_request: None,
+            stream,
+            state: ConnectionState::WaitingForRequestLine,
+            buffer: [0; BUFFER_SIZE],
+            current_read_start: 0,
+            body_vec: vec![],
+            body_bytes_to_be_read: 0,
+            parsed_requests: VecDeque::new(),
+            response_queue: VecDeque::new(),
+            response_buffer: None,
+        }
+    }
+
+    /// Tries to read new bytes from the stream and automatically update the request.
+    /// Meant to be used only with non-blocking streams and an `EPOLL` structure.
+    /// Should be called whenever an `EPOLLIN` event is signaled.
+    pub fn try_read(&mut self) -> Result<(), ConnectionError> {
+        // Read some bytes from the stream, which will be appended to what is already
+        // present in the buffer from a previous call of `try_read`. There are already
+        // `start_index` bytes present in the buffer.
+        let bytes_read = self
+            .stream
+            .read(&mut self.buffer[self.current_read_start..])
+            .map_err(|_| ConnectionError::StreamError)?
+            + self.current_read_start;
+
+        // If the read returned 0 then the client has closed the connection.
+        if bytes_read == self.current_read_start {
+            return Err(ConnectionError::ConnectionClosed);
+        }
+
+        let mut line_start_index = 0;
+        loop {
+            match self.state {
+                ConnectionState::WaitingForRequestLine => {
```

Comment:
The code here is more complex than it needs to be. I think we can use many simplifying assumptions.
The complexity here is added by the fact that we support keeping connections alive. We should discuss whether this is something that we want to support.

Some examples:
- when parsing the request line we should always assume that we have to read from the buffer starting with position=0; this assumption doesn't hold because a new request can be sent part of the same connection hence it will leave in the same buffer.
- the code for parsing the request line would also be much simpler because you always exit with error directly if you don't find the CRLF in the first 1024. In the current code you actually have to check how many bytes you read from the current request in the buffer.

There are some other examples, but we should first decide if we want to keep the connection alive before diving deep into those as well.

CC: @firecracker-microvm/compute-capsule

Reply:
Very few users of Firecracker configure it using a client like ‘curl’ which does one command per connection.

Most users either generate a client from the swagger API spec or use something like ‘firectl’